chronos/asyncloop.nim

```diff
@@ -426,6 +441,159 @@ when defined(windows):
     ## Unregisters ``fd``.
     getThreadDispatcher().handles.excl(fd)
 
+  {.push stackTrace: off.}
+  proc waitableCallback(param: pointer, timerOrWaitFired: WINBOOL) {.
+       stdcall, gcsafe.} =
+    # This procedure will be executed in `wait thread`, so it must not use
+    # GC related objects.
+    # We going to ignore callbacks which was spawned when `isNil(param) == true`
+    # because we unable to indicate this error.
+    if isNil(param):
+      return
+    var wh = cast[WaitableHandle](param)
+    # We ignore result of postQueueCompletionStatus() call because we unable to
+    # indicate error.
+    discard postQueuedCompletionStatus(wh.ioPort, DWORD(timerOrWaitFired),
+                                       ULONG_PTR(wh.handleFd),
+                                       cast[pointer](wh.ovl))
+  {.pop.}
+
+  proc registerWaitable(
+         handle: HANDLE, flags: ULONG,
+         timeout: Duration, cb: CallbackFunc, udata: pointer
+       ): Result[WaitableHandle, OSErrorCode] =
+    ## Register handle of (Change notification, Console input, Event,
+    ## Memory resource notification, Mutex, Process, Semaphore, Thread,
+    ## Waitable timer) for waiting, using specific Windows' ``flags`` and
+    ## ``timeout`` value.
+    ##
+    ## Callback ``cb`` will be scheduled with ``udata`` parameter when
+    ## ``handle`` become signaled.
+    ##
+    ## Result of this procedure call ``WaitableHandle`` should be closed using
+    ## closeWaitable() call.
+    ##
+    ## NOTE: This is private procedure, not supposed to be publicly available,
+    ## please use ``waitForSingleObject()``.
+    let
+      loop = getThreadDispatcher()
+      cell =
+        try:
+          system.protect(rawEnv(cb))
```

Comment:
is this interface documented somewhere? why is it needed, ie what does it do that `GC_ref` doesn't?

Reply:
also, looking at its interface, we would need to call `dispose` as well: https://github.com/nim-lang/Nim/blob/version-1-6/lib/system/gc_common.nim#L96

that said, this does like some cross-thread GC magic, which we're not stressing here yet, from what I can tell?

Reply:
ok, found the calls to dispose, hm

Reply:
system.protect/dispose removed from the code.